test/benchmarks/tcp/xdp.go

```diff
@@ -97,6 +97,14 @@ func newXDPEndpoint(ifaceName string, mac net.HardwareAddr) (stack.LinkEndpoint,
 		return nil, fmt.Errorf("failed to insert socket into BPF map: %v", err)
 	}
 
+	// Ensure that none of the XDP bits are removed by GC'ing the
+	// os.File-wrapped descriptors.
+	for i, fd := range []int{fd, objects.Program.FD(), objects.SockMap.FD(), rawLink.FD()} {
+		if _, err := unix.Dup(fd); err != nil {
+			return nil, fmt.Errorf("failed to dup fd with index %d (%d): %v", i, fd, err)
+		}
+	}
+
```

Comment:
I think you need to add `runtime.KeepAlive(rawLink)` on line 107, otherwise after `rawLink.FD()` returns (but before the FD value is used), the garbage collector can still close the FD.

Reply:
It doesn't matter, as there are now 2 FDs for the file. So whatever the GC does is irrelevant, and it's fine to keep those FDs around as they are meant to live the entire lifetime of the program.

Reply:
Oh, I see your point. Hang on.

Reply:
I am not sure I understand. The `dup(2)` syscall on line 103 will fail if the GC closes `rawLink` before `unix.Dup()` is called. And the failure will be fatal?

Reply:
Done.